### ISSUE TITLE
ossl_dsa_check_key legacy use signature validation

### ISSUE THREAD

**Issue by pm-cfs** [openssl/openssl#19655]
In _providers/common/securitycheck.c_ `ossl_dsa_check_key` calculates L (bits in p) & N (bits in q) and then the comment says it is implementing "The bounds given in SP800 131Ar2 - Table 2".

However, looking at [NIST.SP.800-131Ar2](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-131Ar2.pdf) Table 2, for Legacy use Digital Signature Verification for DSA it says (**_emphasis_** added):
DSA: ((512 ≤ L < 2048) **_or_** (160 ≤ N < 224))

The code in use is:
```
        if (!sign && L < 2048)
            return (L >= 512 && N >= 160 && N < 224);
```

In the if statement, only L is considered, and in the return it is ANDed. But the Table states OR.
This means a 3072 bit DSA cert with SHA1 can't be used for signature verification using the OpenSSL 3 fips provider, and it should be allowed.

My reading is that this should be:
```
        if (!sign && (L < 2048 || N < 224))
            return ((L >= 512 && L < 2048) || (N >= 160 && N < 224));
```

Not sure if anything can be done to correct this without invalidating the FIPS 140-2 Validation Certificate.
If there are other workarounds or if someone has a different reading, I'd love to hear about it.

**Comment by t8m** [comment]
This is deliberate. In fact >=2048 bit L and <224 bit N DSA keys are almost never used.

**Comment by pm-cfs** [comment]
That may be the case "almost never used", but we ran into it with a third party license key issued by government that we can no longer verify in FIPS mode even though by the spec we should be able to do so. We're not at liberty to force them to change their certificate and re-issue around the world.
Legacy is a pain, but forcing others to do better is not always an option.

**Comment by t8m** [comment]
@slontis @paulidale Could this be acceptable change for 3.1?

**Comment by slontis** [comment]
Yes. This looks like a error on my part. I assume it is an OR because there are still use cases out there.. 
If a patch is ever done for 3.0 it should probably be changed there also.

If you are stuck with this, then the only thing you could do in the interim to 'work around' it would be to turn off "security_checks" in the fips config file (Not something I would generally advise to do). See the documentation related to fipsinstall for how to do this.

**Comment by paulidale** [comment]
Looks like a bug to me, so 3.1 and 3.0 would be suitable.

**Comment by t8m** [comment]
Given DSA is really a legacy algorithm, I am moving this to `inactive` state. This is low priority and will get closed at the end of the 3.4.0 release cycle.
